stan/math/rev/functor/cvodes_integrator_adjoint.hpp

```diff
@@ -0,0 +1,787 @@
+    save_varis(args_varis_, args...);
+
+    check_finite(function_name, "initial state", y0);
+    check_finite(function_name, "initial time", t0);
+    check_finite(function_name, "times", ts);
+
+    stan::math::for_each(
+        [func_name = function_name](auto&& arg) {
+          check_finite(func_name, "ode parameters and data", arg);
+        },
+        local_args_tuple_);
+
+    check_nonzero_size(function_name, "times", ts);
+    check_nonzero_size(function_name, "initial state", y0);
+    check_sorted(function_name, "times", ts);
+    check_less(function_name, "initial time", t0, ts[0]);
+    check_positive_finite(function_name, "relative_tolerance_forward",
+                          relative_tolerance_forward_);
+    check_positive_finite(function_name, "absolute_tolerance_forward",
+                          absolute_tolerance_forward_);
+    check_size_match(function_name, "absolute_tolerance_forward",
+                     absolute_tolerance_forward_.size(), "states", N_);
+    check_positive_finite(function_name, "relative_tolerance_backward",
+                          relative_tolerance_backward_);
+    check_positive_finite(function_name, "absolute_tolerance_backward",
+                          absolute_tolerance_backward_);
+    check_size_match(function_name, "absolute_tolerance_backward",
+                     absolute_tolerance_backward_.size(), "states", N_);
+    check_positive_finite(function_name, "relative_tolerance_quadrature",
+                          relative_tolerance_quadrature_);
+    check_positive_finite(function_name, "absolute_tolerance_quadrature",
+                          absolute_tolerance_quadrature_);
+    check_positive(function_name, "max_num_steps", max_num_steps_);
+    check_positive(function_name, "num_steps_between_checkpoints",
+                   num_steps_between_checkpoints_);
+    // for polynomial: 1=CV_HERMITE / 2=CV_POLYNOMIAL
+    check_bounded(function_name, "interpolation_polynomial",
+                  interpolation_polynomial_, 1, 2);
+    // 1=Adams=CV_ADAMS, 2=BDF=CV_BDF
+    check_bounded(function_name, "solver_forward", solver_forward_, 1, 2);
+    check_bounded(function_name, "solver_backward", solver_backward_, 1, 2);
+
+    solver_ = new cvodes_solver(
+        function_name, f, N_, num_args_vars_, ts_.size(), solver_forward_,
+        state_forward_, state_backward_, quad_, absolute_tolerance_forward_,
+        absolute_tolerance_backward_);
+
+    check_flag_sundials(
+        CVodeInit(solver_->cvodes_mem_, &cvodes_integrator_adjoint_vari::cv_rhs,
+                  value_of(t0_), solver_->nv_state_forward_),
+        "CVodeInit");
+
+    // Assign pointer to this as user data
+    check_flag_sundials(
+        CVodeSetUserData(solver_->cvodes_mem_, reinterpret_cast<void*>(this)),
+        "CVodeSetUserData");
+
+    cvodes_set_options(solver_->cvodes_mem_, relative_tolerance_forward_,
+                       absolute_tolerance_forward_(0), max_num_steps_);
```

Comment:
It should probably be that

```cpp
// Initialize solver parameters
check_flag_sundials(CVodeSStolerances(cvodes_mem, rel_tol, abs_tol),
                      "CVodeSStolerances");
```

moves out of `covdes_set_options`. This requires a bit of copy-paste with the other solvers, but it seems awkward to set the forward tolerances like this and then overwrite them.

Reply:
ok.